source/common/quic/platform/quiche_flags_impl.h

```diff
@@ -6,128 +6,17 @@
-template <typename T> class TypedFlag : public Flag {
-public:
-  TypedFlag(const char* name, T default_value, const char* help)
-      : Flag(name, help), value_(default_value), default_value_(default_value) {}
-
-  bool setValueFromString(const std::string& value_str) override;
-
-  void resetValue() override {
-    absl::MutexLock lock(&mutex_);
-    value_ = default_value_;
-  }
-
-  // Set flag value.
-  void setValue(T value) {
-    absl::MutexLock lock(&mutex_);
-    value_ = value;
-  }
-
-  // Return flag value.
-  T value() const {
-    absl::MutexLock lock(&mutex_);
-    return value_;
-  }
-
-private:
-  mutable absl::Mutex mutex_;
-  T value_ ABSL_GUARDED_BY(mutex_);
-  T default_value_;
-};
-
-// SetValueFromString specializations
-template <> bool TypedFlag<bool>::setValueFromString(const std::string& value_str);
-template <> bool TypedFlag<int32_t>::setValueFromString(const std::string& value_str);
-template <> bool TypedFlag<int64_t>::setValueFromString(const std::string& value_str);
-template <> bool TypedFlag<double>::setValueFromString(const std::string& value_str);
-template <> bool TypedFlag<std::string>::setValueFromString(const std::string& value_str);
-template <> bool TypedFlag<unsigned long>::setValueFromString(const std::string& value_str);
-template <> bool TypedFlag<unsigned long long>::setValueFromString(const std::string& value_str);
-
-// Flag declarations
-#define QUIC_FLAG(flag, ...) extern TypedFlag<bool>* flag;
-#include "quiche/quic/core/quic_flags_list.h"
-QUIC_FLAG(FLAGS_quic_reloadable_flag_spdy_testonly_default_false, false)
-QUIC_FLAG(FLAGS_quic_reloadable_flag_spdy_testonly_default_true, true)
-QUIC_FLAG(FLAGS_quic_restart_flag_spdy_testonly_default_false, false)
-QUIC_FLAG(FLAGS_quic_restart_flag_spdy_testonly_default_true, true)
-QUIC_FLAG(FLAGS_quic_reloadable_flag_http2_testonly_default_false, false)
-QUIC_FLAG(FLAGS_quic_reloadable_flag_http2_testonly_default_true, true)
-QUIC_FLAG(FLAGS_quic_restart_flag_http2_testonly_default_false, false)
-QUIC_FLAG(FLAGS_quic_restart_flag_http2_testonly_default_true, true)
-#undef QUIC_FLAG
-
-#define QUIC_PROTOCOL_FLAG(type, flag, ...) extern TypedFlag<type>* FLAGS_##flag;
-#include "quiche/quic/core/quic_protocol_flags_list.h"
-#undef QUIC_PROTOCOL_FLAG
+#include "common/quic/platform/flags_impl.h"
+#include "common/runtime/runtime_features.h"
 
 // |flag| is the global flag variable, which is a pointer to TypedFlag<type>.
 #define GetQuicheFlagImpl(flag) (quiche::flag)->value()
 
 // |flag| is the global flag variable, which is a pointer to TypedFlag<type>.
 #define SetQuicheFlagImpl(flag, value) (quiche::flag)->setValue(value)
 
-#define GetQuicheReloadableFlagImpl(module, flag) quiche::FLAGS_quic_reloadable_flag_##flag->value()
+#define GetQuicheReloadableFlagImpl(module, flag)                                                  \
+  Envoy::Runtime::runtimeFeatureEnabled(EnvoyReloadableFeature(flag))
```

Comment:
> The GetQuicheReloadableFlagImpl implementation seems inconsistent with SetQuicheReloadableFlagImpl below.
> 
> What happens if some code accesses the FLAGS_quic_reloadable_flag_ directly in quiche code instead of calling GetQuicheReloadableFlagImpl?
> 
> It seems to me that SetQuicheReloadableFlagImpl is only usable in tests, and only if there's no runtime feature override for the flag in question.

SetQuicheReloadableFlagImpl() is used in non-test code now: SetQuicReloadableFlag(quic_single_ack_in_packet2, true);
This is to apply a fix in QUICHE code only affects Envoy.

> 
> Also, note that runtimeFeatureEnabled is significantly more costly than accessing FLAGS_quic_reloadable_flags. Can we make sure that there are no GetQuicheReloadableFlagImpl calls in tight loops?

We can't guarantee that as QUICHE latches a flag value based on behavior consistency not performance, as GetQuicheReloadableFlagImpl() isn't costly in our implementation.

Reply:
As Dan mentioned, it's very likely for Envoy to have different default values for some flags compared to core Quiche.
Today in the gfe flag file, we have a field "chrome_value:" to set special default value in chrome. I imagine we can have similar things for Envoy in the future.

As for the current Envoy setup, my view is that the quic_flags_list.cc will be the source of truth, and SetQuicReloadableFlag() will modify the source values.
However, runtime override has the highest priority. I added a test runtime_impl_test for this.

Reply:
As Dan mentioned, the flag based implementation of the Get method is significantly more performant than the runtime feature variant.  QUICHE doesn't always latch values of reloadable flags as it assumes that Get is fast.

My question is how can we propagate state specified via runtime features to QUIC reloadable flags?  An option that we may want to consider is copying the value computed from runtime feature to the boolean flag representations.

Reply:
Hmm great point!
Having QUIC code calling into runtime feature in a tight loop does degrade performance.

We could potentially cache the calculated runtime value or even set the source of truth with runtime override. But either way we will need some sort of notification mechanism. How hard do you think it will be to hook this up?

Currently in the QUIC codebase, I don't think there's any Get call in tight loop, and multiple Get calls for a single flags in an object are strongly encouraged to use a latched value. So if the notification system is hard to set up, we could mitigate the runtime feature complexity by enforcing flag latching in Quiche.

Reply:
I'm not familiar with the code that reloads runtime flags, but I think that introducing a notification mechanism on runtime feature changes should be straightforward.  Should I try to do some research on it and send you some pointers?